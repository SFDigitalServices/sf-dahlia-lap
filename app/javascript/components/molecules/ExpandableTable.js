import React, { useState, useEffect } from 'react'
import classNames from 'classnames'
import { kebabCase } from 'lodash'

import formUtils from '~/utils/formUtils'

const ExpandableTableRow = ({
  row,
  rowKeyIndex,
  numColumns,
  expanderRenderer,
  expandedRowRenderer,
  original,
  closeRow = false
}) => {
  const [expanded, setExpanded] = useState(false)

  useEffect(() => {
    if (closeRow) {
      setExpanded(false)
    }
  }, [closeRow])

  const toggleExpandedRow = () => setExpanded(!expanded)

  const cells = row.map((datum, j) => (
    <td className={datum.classes ? datum.classes.join(' ') : ''} key={j}>
      {datum.formatType === 'currency' ? formUtils.formatPrice(datum.content) : datum.content}
    </td>
  ))

  const rowId = rowKeyIndex ? kebabCase(row[rowKeyIndex].content) : null

  return (
    <>
      <tr className='tr-expand' aria-expanded={expanded} id={rowId ? `${rowId}-row` : null}>
        {cells}
        <td key='expander'>
          {expanderRenderer && expanderRenderer(row, expanded, toggleExpandedRow)}
        </td>
      </tr>
      <tr
        className='tr-expand-content'
        aria-hidden={!expanded}
        id={rowId ? `${rowId}-panel` : null}
      >
        <td colSpan={numColumns} className='td-expand-nested no-padding'>
          {expandedRowRenderer && expandedRowRenderer(row, toggleExpandedRow, original)}
        </td>
      </tr>
    </>
  )
}

const ExpandableTable = ({
  columns,
  rows,
  rowKeyIndex,
  expanderRenderer,
  expandedRowRenderer,
  originals,
  classes,
  closeAllRows
}) => (
  <table
    className={classNames('td-light td-plain th-plain', classes)}
    style={{ background: 'transparent' }}
    role='grid'
  >
    <thead>
      <tr>
        {columns.map((column, i) => (
          <th key={i} scope='col' className={column.classes ? column.classes.join(' ') : ''}>
            {column.content}
          </th>
        ))}
      </tr>
    </thead>
    <tbody>
      {rows.map((row, i) => (
        <ExpandableTableRow
          key={i}
          rowKeyIndex={rowKeyIndex} // index for element in row to use as a css id.
          idx={i}
          original={originals && originals[i]}
          row={row}
          numColumns={columns.length}
          expanderRenderer={expanderRenderer}
          expandedRowRenderer={expandedRowRenderer}
          closeRow={closeAllRows}
        />
      ))}
    </tbody>
  </table>
)

ExpandableTable.ExpanderButton = ({ onClick, label = 'Expand', id }) => (
  <button type='button' className='button button-link action-link' onClick={onClick} id={id}>
    {label}
  </button>
)

export default ExpandableTable
