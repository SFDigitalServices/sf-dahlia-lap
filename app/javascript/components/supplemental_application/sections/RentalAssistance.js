import React, { useEffect, useState } from 'react'

import { isEmpty, findIndex } from 'lodash'

import Button from 'components/atoms/Button'
import TableWrapper from 'components/atoms/TableWrapper'
import ExpandableTable from 'components/molecules/ExpandableTable'
import FormGrid from 'components/molecules/FormGrid'
import InlineModal from 'components/molecules/InlineModal'
import {
  CurrencyField,
  HelpText,
  InputField,
  SelectField,
  YesNoRadioGroup
} from 'utils/form/final_form/Field'
import { isSingleRentalAssistanceValid } from 'utils/form/formSectionValidations'
import validate, { convertCurrency } from 'utils/form/validations'
import formUtils from 'utils/formUtils'

import { withContext } from '../context'

const { ExpanderButton } = ExpandableTable

const typeOfAssistance = [
  'Catholic Charities',
  'Compass Family',
  'Glide',
  'Hamilton Family',
  'Homeless Prenatal',
  'HOPWA',
  'Q Foundation: First Monthâ€™s Rent',
  'Q Foundation: Move-in Deposit',
  'Q Foundation: Person with Disability Rent Subsidy',
  'Q Foundation: Senior Rent Subsidy',
  'San Francisco AIDS Foundation',
  'Section 8 HCV (tenant) voucher',
  'Self-Help for the Elderly',
  'VASH voucher',
  'Other'
]

const typeOfAssistanceOptions = formUtils.toOptions(typeOfAssistance)

const isOther = (values) => values && values.type_of_assistance === 'Other'

export const RentalAssistanceTable = ({
  form,
  rentalAssistances,
  onCancelEdit,
  onSave,
  onDelete,
  applicationMembers,
  disabled,
  loading
}) => {
  const columns = [
    { content: 'Recipient' },
    { content: 'Type' },
    { content: 'Amount' },
    { content: 'Recurring' },
    { content: '' }
  ]

  const expanderRenderer = (row, expanded, expandedRowToggler) => {
    return !expanded && !disabled && <ExpanderButton label='Edit' onClick={expandedRowToggler} />
  }

  const expandedRowRenderer = (rentalAssistances, form) => (row, toggle, original) => {
    const index = findIndex(rentalAssistances, original)

    // Run the async function, if the result doesn't throw an error and returned a
    // truthy response, toggle the panel to show/hide
    const toggleIfSuccessful = (asyncFunc) => (...args) =>
      asyncFunc(...args).then((result) => {
        if (result) toggle()
      })

    const handleClose = () => {
      onCancelEdit(index)
      toggle()
    }

    const handleDelete = () => toggleIfSuccessful(onDelete)(rentalAssistances[index])
    const handleSave = (index) => toggleIfSuccessful(onSave)(index)

    return (
      <RentalAssistanceForm
        values={original}
        row={row}
        onSave={handleSave}
        onClose={handleClose}
        onDelete={handleDelete}
        index={index}
        applicationMembers={applicationMembers}
        loading={loading}
        form={form}
      />
    )
  }

  const buildRows = () =>
    rentalAssistances.map((ra) => {
      const appMember = applicationMembers.find((m) => m.id === ra.recipient)
      const appMemberName = appMember ? `${appMember.first_name} ${appMember.last_name}` : ''

      return [
        { content: appMemberName },
        { content: ra.other_assistance_name || ra.type_of_assistance },
        { content: ra.assistance_amount ? ra.assistance_amount : '', formatType: 'currency' },
        { content: ra.recurring_assistance }
      ]
    })

  const rows = buildRows()

  return (
    <TableWrapper marginTop>
      <ExpandableTable
        originals={rentalAssistances}
        columns={columns}
        rows={rows}
        expanderRenderer={expanderRenderer}
        expandedRowRenderer={expandedRowRenderer(rentalAssistances, form)}
        closeAllRows={disabled}
        classes={['rental-assistances']}
      />
    </TableWrapper>
  )
}

export const RentalAssistanceForm = ({
  values,
  onSave,
  loading,
  onClose,
  applicationMembers,
  onDelete,
  isNew,
  index,
  form,
  visited
}) => {
  const validateAssistanceAmount = (value) => {
    return (
      validate.isValidCurrency('Please enter a valid dollar amount.')(value) ||
      validate.isUnderMaxValue(Math.pow(10, 5))('Please enter a smaller number.')(value)
    )
  }

  const isFormValid = () => {
    const isValid = isSingleRentalAssistanceValid(form, index)
    if (!isValid) {
      // Force submit to show errors on forms
      form.submit()
    }
    return isValid
  }

  const applicationMembersOptions = applicationMembers.map((member) => ({
    label: `${member.first_name} ${member.last_name}`,
    value: member.id
  }))

  const handleSave = () => {
    if (isFormValid()) {
      onSave(index)
    }
  }

  const getField = (assistanceField) => `rental_assistances.${index}.${assistanceField}`

  const modalId = isNew ? 'rental-assistance-new-form' : `rental-assistance-edit-form-${index}`

  return (
    <InlineModal whiteBackground marginBottom={isNew} id={modalId}>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            label='Recipient'
            fieldName={getField('recipient')}
            options={applicationMembersOptions}
            className='rental-assistance-recipient'
          />
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row>
        <FormGrid.Item>
          <SelectField
            label='Type of Assistance'
            fieldName={getField('type_of_assistance')}
            options={typeOfAssistanceOptions}
            className='rental-assistance-type'
            validation={validate.isPresent('Please select a type of assistance.')}
          />
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row>
        <FormGrid.Item>
          <CurrencyField
            label='Assistance Amount'
            fieldName={getField('assistance_amount')}
            validation={validateAssistanceAmount}
            id='assistance_amount'
            isDirty={visited && visited[getField('assistance_amount')]}
          />
        </FormGrid.Item>
      </FormGrid.Row>
      <FormGrid.Row>
        <FormGrid.Item>
          <YesNoRadioGroup
            label='Recurring Assistance'
            fieldName={getField('recurring_assistance')}
            uniqId={values?.id || 'new'}
            trueValue='Yes'
            falseValue='No'
            className='rental-assistance-recurring'
          />
        </FormGrid.Item>
      </FormGrid.Row>
      {isOther(form.getState().values.rental_assistances[index]) && (
        <FormGrid.Row>
          <FormGrid.Item>
            <InputField
              label='Other Assistance Name'
              fieldName={getField('other_assistance_name')}
            />
          </FormGrid.Item>
        </FormGrid.Row>
      )}
      <FormGrid.Row>
        <div className='form-grid_item column'>
          <Button
            classes='primary margin-right'
            tiny
            onClick={handleSave}
            disabled={loading}
            noBottomMargin
            text={loading ? 'Saving...' : 'Save'}
            id='rental-assistance-save'
          />
          <Button
            classes='secondary'
            tiny
            onClick={onClose}
            disabled={loading}
            noBottomMargin
            text='Cancel'
            id='rental-assistance-cancel'
          />
          {!isNew && (
            <Button
              classes='alert-fill right'
              tiny
              onClick={onDelete}
              disabled={loading}
              noBottomMargin
              text='Delete'
              id='rental-assistance-delete'
            />
          )}
        </div>
      </FormGrid.Row>
    </InlineModal>
  )
}

const RentalAssistance = ({ store, form, visited, disabled }) => {
  const {
    application,
    applicationMembers,
    handleDeleteRentalAssistance,
    handleSaveRentalAssistance,
    loading
  } = store

  const [isEditingNewAssistance, setIsEditingNewAssistance] = useState(false)

  /**
   * When the disabled prop changes to false, hide the new assistance panel.
   */
  useEffect(() => {
    if (disabled) {
      setIsEditingNewAssistance(false)
    }
  }, [disabled])

  const handleOpenNewPanel = () => {
    setIsEditingNewAssistance(true)
  }

  const handleCloseNewPanel = () => {
    setIsEditingNewAssistance(false)
    handleCancelEdit(application.rental_assistances.length)
  }

  const handleCancelEdit = (index) => {
    const isNewAssistance = index === application.rental_assistances.length
    const assistanceToRevertTo = isNewAssistance ? {} : application.rental_assistances[index]
    form.change(`rental_assistances.${index}`, assistanceToRevertTo)
  }

  const handleSave = async (index, action = 'update') => {
    const entireForm = form.getState().values
    const rentalAssistance = convertCurrency(entireForm.rental_assistances[index])
    return handleSaveRentalAssistance(rentalAssistance, entireForm, action)
  }

  const handleSaveNewAssistance = async (index) =>
    handleSave(index, 'create').then(() => setIsEditingNewAssistance(false))

  const handleDelete = async (rentalAssistance) =>
    handleDeleteRentalAssistance(rentalAssistance, form.getState().values)

  return (
    <>
      {!isEmpty(application.rental_assistances) && (
        <RentalAssistanceTable
          rentalAssistances={application.rental_assistances}
          applicationMembers={applicationMembers}
          onCancelEdit={handleCancelEdit}
          onDelete={handleDelete}
          onSave={handleSave}
          form={form}
          loading={loading}
          disabled={disabled}
        />
      )}

      {!disabled && isEditingNewAssistance && (
        <FormGrid.Row expand={false}>
          <RentalAssistanceForm
            onSave={handleSaveNewAssistance}
            onClose={handleCloseNewPanel}
            applicationMembers={applicationMembers}
            loading={loading}
            values={{ type_of_assistance: null }}
            index={application.rental_assistances.length}
            form={form}
            visited={visited}
            isNew
          />
        </FormGrid.Row>
      )}
      {!isEditingNewAssistance && (
        <FormGrid.Row>
          <FormGrid.Item>
            {!disabled && (
              <Button
                id='add-rental-assistance'
                text='Add Rental Assistance'
                small
                disabled={loading}
                onClick={handleOpenNewPanel}
              />
            )}
            <div className={disabled ? 'margin-top' : 'margin-top--half'} />
            <HelpText note='Rental Assistance includes recurring vouchers and subsidies, as well as one-time grants and other assistance.' />
          </FormGrid.Item>
        </FormGrid.Row>
      )}
    </>
  )
}

export default withContext(RentalAssistance)
