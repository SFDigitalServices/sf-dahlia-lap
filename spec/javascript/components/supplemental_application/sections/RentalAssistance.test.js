import React from 'react'
import { cloneDeep } from 'lodash'
import RentalAssistance, {
  RentalAssistanceForm,
  RentalAssistanceTable
} from '~/components/supplemental_application/sections/RentalAssistance'
import {
  withForm,
  shallowWithFormAndContext,
  findByNameAndProps
} from '../../../testUtils/wrapperUtil'
import {
  InputField
} from '~/utils/form/final_form/Field'
import ExpandableTable from '~/components/molecules/ExpandableTable'

const baseContext = {
  application: { rental_assistances: [] },
  applicationMembers: [{ id: '123', first_name: 'Test', last_name: 'Tester' }]
}

const rentalAssistance = {
  type_of_assistance: 'Compass Family',
  recurring_assistance: 'Yes',
  assistance_amount: '100',
  recipient: '123'
}

describe('RentalAssistance', () => {
  const getWrapper = (context) => shallowWithFormAndContext(
    context,
    (form) => <RentalAssistance form={form} />
  )

  test('matches snapshot', () => {
    const context = cloneDeep(baseContext)
    context.application.rental_assistances = [rentalAssistance]
    const wrapper = getWrapper(context)
    expect(wrapper).toMatchSnapshot()
  })

  test('should not render a table if rental assistances is empty', () => {
    const context = cloneDeep(baseContext)
    const wrapper = getWrapper(context)
    expect(wrapper.find(RentalAssistanceTable)).toHaveLength(0)
  })

  test('should render a table if rental assistances are present', () => {
    const context = cloneDeep(baseContext)
    context.application.rental_assistances = [rentalAssistance]
    const wrapper = getWrapper(context)
    expect(wrapper.find(RentalAssistanceTable)).toHaveLength(1)
  })

  test('should render the Add Rental Assistance button by default', () => {
    const context = cloneDeep(baseContext)
    const wrapper = getWrapper(context)
    expect(findByNameAndProps(wrapper, 'Button', { text: 'Add Rental Assistance' })).toHaveLength(1)
  })

  test('does not render the create new assistance form by default', () => {
    const context = cloneDeep(baseContext)
    const wrapper = getWrapper(context)
    expect(wrapper.find(RentalAssistanceForm)).toHaveLength(0)
  })

  describe('after the rental assistance button is clicked', () => {
    let wrapper

    beforeEach(() => {
      wrapper = getWrapper(cloneDeep(baseContext))
      findByNameAndProps(wrapper, 'Button', { text: 'Add Rental Assistance' }).simulate('click')
    })

    test('should render the new rental assistance form after the add rental assistance button is clicked', () => {
      expect(wrapper.find(RentalAssistanceForm)).toHaveLength(1)
    })

    test('should hide the new rental assistance form after cancel is clicked', () => {
      wrapper.find(RentalAssistanceForm).prop('onClose')()
      expect(wrapper.find(RentalAssistanceForm)).toHaveLength(0)
      expect(findByNameAndProps(wrapper, 'Button', { text: 'Add Rental Assistance' })).toHaveLength(1)
    })
  })
})

describe('RentalAssistanceTable', () => {
  const getWrapper = ({ disabled = false }) => {
    const context = cloneDeep(baseContext)
    context.application.rental_assistances = [rentalAssistance]

    return withForm(
      context,
      (form) => (
        <RentalAssistanceTable
          form={form}
          rentalAssistances={[rentalAssistance]}
          applicationMembers={[]}
          disabled={disabled}
        />
      )
    )
  }

  test('should render an ExpanderTable', () => {
    const wrapper = getWrapper({})
    expect(wrapper.find(ExpandableTable)).toHaveLength(1)
  })

  test('should not close all panels if not disabled', () => {
    const wrapper = getWrapper({})
    expect(wrapper.find(ExpandableTable).prop('closeAllRows')).toBeFalsy()
  })

  test('should close all panels when disabled', () => {
    const wrapper = getWrapper({ disabled: true })
    expect(wrapper.find(ExpandableTable).prop('closeAllRows')).toBeTruthy()
  })
})

describe('RentalAssistanceForm', () => {
  const getWrapper = ({
    assistance,
    isNew = false,
    onSave = () => {},
    onClose = () => {},
    onDelete = () => {},
    shouldMount = false
  }) => {
    const context = cloneDeep(baseContext)
    context.application.rental_assistances = assistance ? [assistance] : []

    return withForm(
      context.application,
      (form) => (
        <RentalAssistanceForm
          values={assistance}
          index={0}
          applicationMembers={context.applicationMembers}
          form={form}
          isNew={isNew}
          onSave={onSave}
          onClose={onClose}
          onDelete={onDelete}
        />
      ),
      shouldMount
    )
  }

  test('should not show the Other Assistance Name input when Type of Assistance is not Other', () => {
    const wrapper = getWrapper({ assistance: rentalAssistance })
    expect(wrapper.find(InputField)).toHaveLength(0)
  })

  test('should show the Other Assistance Name when Type of Assistance is Other', () => {
    const wrapper = getWrapper({
      assistance: {
        ...rentalAssistance,
        type_of_assistance: 'Other',
        other_assistance_name: 'Test Other Assistance Name'
      }
    })
    expect(wrapper.find(InputField)).toHaveLength(1)
    expect(wrapper.find(InputField).prop('label')).toEqual('Other Assistance Name')
  })

  test('should validate that type of assistance is present upon save of the form panel', () => {
    // need to mount it to access the error classes
    const wrapper = getWrapper({ assistance: null, shouldMount: true })

    findByNameAndProps(wrapper, 'Button', { text: 'Save' }).simulate('click')
    expect(wrapper.find('.rental-assistance-type.error').exists()).toBeTruthy()
  })

  describe('when it is new', () => {
    let mockSaveCallback
    let mockCloseCallback
    let mockDeleteCallback
    let wrapper

    beforeEach(() => {
      mockSaveCallback = jest.fn()
      mockCloseCallback = jest.fn()
      mockDeleteCallback = jest.fn()

      wrapper = getWrapper({
        assistance: rentalAssistance,
        isNew: true,
        onSave: mockSaveCallback,
        onClose: mockCloseCallback,
        onDelete: mockDeleteCallback
      })
    })

    test('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    test('should render the correct modal ID', () => {
      expect(wrapper.find('#rental-assistance-new-form')).toHaveLength(1)
    })

    test('should not render the delete button', () => {
      expect(findByNameAndProps(wrapper, 'Button', { text: 'Delete' })).toHaveLength(0)
    })

    test('should call the save callback when clicked', () => {
      findByNameAndProps(wrapper, 'SelectField', { label: 'Type of Assistance' })
        .simulate('change', { target: { value: 'Catholic Charities' } })

      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
      findByNameAndProps(wrapper, 'Button', { text: 'Save' }).simulate('click')
      expect(mockSaveCallback.mock.calls).toHaveLength(1)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
    })

    test('should call the cancel callback when clicked', () => {
      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
      findByNameAndProps(wrapper, 'Button', { text: 'Cancel' }).simulate('click')
      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(1)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
    })
  })

  describe('when it already exists on salesforce', () => {
    let mockSaveCallback
    let mockCloseCallback
    let mockDeleteCallback
    let wrapper

    beforeEach(() => {
      mockSaveCallback = jest.fn()
      mockCloseCallback = jest.fn()
      mockDeleteCallback = jest.fn()

      wrapper = getWrapper({
        assistance: rentalAssistance,
        onSave: mockSaveCallback,
        onClose: mockCloseCallback,
        onDelete: mockDeleteCallback
      })
    })

    test('matches snapshot', () => {
      expect(wrapper).toMatchSnapshot()
    })

    test('should render the correct modal ID', () => {
      expect(wrapper.find('#rental-assistance-edit-form-0')).toHaveLength(1)
    })

    test('should render the delete button', () => {
      expect(findByNameAndProps(wrapper, 'Button', { text: 'Delete' })).toHaveLength(1)
    })

    test('should call the save callback when clicked', () => {
      findByNameAndProps(wrapper, 'SelectField', { label: 'Type of Assistance' })
        .simulate('change', { target: { value: 'Catholic Charities' } })

      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
      findByNameAndProps(wrapper, 'Button', { text: 'Save' }).simulate('click')
      expect(mockSaveCallback.mock.calls).toHaveLength(1)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
    })

    test('should call the delete callback when clicked', () => {
      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
      findByNameAndProps(wrapper, 'Button', { text: 'Delete' }).simulate('click')
      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(1)
    })

    test('should call the cancel callback when clicked', () => {
      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(0)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
      findByNameAndProps(wrapper, 'Button', { text: 'Cancel' }).simulate('click')
      expect(mockSaveCallback.mock.calls).toHaveLength(0)
      expect(mockCloseCallback.mock.calls).toHaveLength(1)
      expect(mockDeleteCallback.mock.calls).toHaveLength(0)
    })
  })
})
