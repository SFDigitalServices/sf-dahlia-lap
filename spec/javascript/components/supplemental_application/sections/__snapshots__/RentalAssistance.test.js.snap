// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RentalAssistance matches snapshot 1`] = `
<Fragment>
  <RentalAssistanceTable
    applicationMembers={
      Array [
        Object {
          "first_name": "Test",
          "id": "123",
          "last_name": "Tester",
        },
      ]
    }
    form={
      Object {
        "batch": [Function],
        "blur": [Function],
        "change": [Function],
        "destroyOnUnregister": false,
        "focus": [Function],
        "getFieldState": [Function],
        "getRegisteredFields": [Function],
        "getState": [Function],
        "initialize": [Function],
        "isValidationPaused": [Function],
        "mutators": Object {},
        "pauseValidation": [Function],
        "registerField": [Function],
        "reset": [Function],
        "resetFieldState": [Function],
        "restart": [Function],
        "resumeValidation": [Function],
        "setConfig": [Function],
        "submit": [Function],
        "subscribe": [Function],
      }
    }
    onCancelEdit={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
    rentalAssistances={
      Array [
        Object {
          "assistance_amount": "100",
          "recipient": "123",
          "recurring_assistance": "Yes",
          "type_of_assistance": "Compass Family",
        },
      ]
    }
  />
  <Component
    expand={true}
  >
    <Component>
      <Button
        id="add-rental-assistance"
        onClick={[Function]}
        small={true}
        text="Add Rental Assistance"
      />
      <div
        className="margin-top--half"
      />
      <HelpText
        note="Rental Assistance includes recurring vouchers and subsidies, as well as one-time grants and other assistance."
      />
    </Component>
  </Component>
</Fragment>
`;

exports[`RentalAssistanceForm when it already exists on salesforce matches snapshot 1`] = `
<InlineModal
  id="rental-assistance-edit-form-0"
  marginBottom={false}
  whiteBackground={true}
>
  <Component
    expand={true}
  >
    <Component>
      <SelectField
        className="rental-assistance-recipient"
        fieldName="rental_assistances.0.recipient"
        label="Recipient"
        options={
          Array [
            Object {
              "label": "Test Tester",
              "value": "123",
            },
          ]
        }
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <Component>
      <SelectField
        className="rental-assistance-type"
        fieldName="rental_assistances.0.type_of_assistance"
        label="Type of Assistance"
        options={
          Array [
            Object {
              "label": "Catholic Charities",
              "value": "Catholic Charities",
            },
            Object {
              "label": "Compass Family",
              "value": "Compass Family",
            },
            Object {
              "label": "Glide",
              "value": "Glide",
            },
            Object {
              "label": "Hamilton Family",
              "value": "Hamilton Family",
            },
            Object {
              "label": "Homeless Prenatal",
              "value": "Homeless Prenatal",
            },
            Object {
              "label": "HOPWA",
              "value": "HOPWA",
            },
            Object {
              "label": "Q Foundation: First Month’s Rent",
              "value": "Q Foundation: First Month’s Rent",
            },
            Object {
              "label": "Q Foundation: Move-in Deposit",
              "value": "Q Foundation: Move-in Deposit",
            },
            Object {
              "label": "Q Foundation: Person with Disability Rent Subsidy",
              "value": "Q Foundation: Person with Disability Rent Subsidy",
            },
            Object {
              "label": "Q Foundation: Senior Rent Subsidy",
              "value": "Q Foundation: Senior Rent Subsidy",
            },
            Object {
              "label": "San Francisco AIDS Foundation",
              "value": "San Francisco AIDS Foundation",
            },
            Object {
              "label": "Section 8 HCV (tenant) voucher",
              "value": "Section 8 HCV (tenant) voucher",
            },
            Object {
              "label": "Self-Help for the Elderly",
              "value": "Self-Help for the Elderly",
            },
            Object {
              "label": "VASH voucher",
              "value": "VASH voucher",
            },
            Object {
              "label": "Other",
              "value": "Other",
            },
          ]
        }
        validation={[Function]}
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <Component>
      <CurrencyField
        fieldName="rental_assistances.0.assistance_amount"
        id="assistance_amount"
        label="Assistance Amount"
        validation={[Function]}
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <Component>
      <YesNoRadioGroup
        className="rental-assistance-recurring"
        falseValue="No"
        fieldName="rental_assistances.0.recurring_assistance"
        label="Recurring Assistance"
        trueValue="Yes"
        uniqId="new"
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <div
      className="form-grid_item column"
    >
      <Button
        classes="primary margin-right"
        disabled={false}
        id="rental-assistance-save"
        noBottomMargin={true}
        onClick={[Function]}
        text="Save"
        tiny={true}
      />
      <Button
        classes="secondary"
        disabled={false}
        id="rental-assistance-cancel"
        noBottomMargin={true}
        onClick={[MockFunction]}
        text="Cancel"
        tiny={true}
      />
      <Button
        classes="alert-fill right"
        disabled={false}
        id="rental-assistance-delete"
        noBottomMargin={true}
        onClick={[MockFunction]}
        text="Delete"
        tiny={true}
      />
    </div>
  </Component>
</InlineModal>
`;

exports[`RentalAssistanceForm when it is new matches snapshot 1`] = `
<InlineModal
  id="rental-assistance-new-form"
  marginBottom={true}
  whiteBackground={true}
>
  <Component
    expand={true}
  >
    <Component>
      <SelectField
        className="rental-assistance-recipient"
        fieldName="rental_assistances.0.recipient"
        label="Recipient"
        options={
          Array [
            Object {
              "label": "Test Tester",
              "value": "123",
            },
          ]
        }
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <Component>
      <SelectField
        className="rental-assistance-type"
        fieldName="rental_assistances.0.type_of_assistance"
        label="Type of Assistance"
        options={
          Array [
            Object {
              "label": "Catholic Charities",
              "value": "Catholic Charities",
            },
            Object {
              "label": "Compass Family",
              "value": "Compass Family",
            },
            Object {
              "label": "Glide",
              "value": "Glide",
            },
            Object {
              "label": "Hamilton Family",
              "value": "Hamilton Family",
            },
            Object {
              "label": "Homeless Prenatal",
              "value": "Homeless Prenatal",
            },
            Object {
              "label": "HOPWA",
              "value": "HOPWA",
            },
            Object {
              "label": "Q Foundation: First Month’s Rent",
              "value": "Q Foundation: First Month’s Rent",
            },
            Object {
              "label": "Q Foundation: Move-in Deposit",
              "value": "Q Foundation: Move-in Deposit",
            },
            Object {
              "label": "Q Foundation: Person with Disability Rent Subsidy",
              "value": "Q Foundation: Person with Disability Rent Subsidy",
            },
            Object {
              "label": "Q Foundation: Senior Rent Subsidy",
              "value": "Q Foundation: Senior Rent Subsidy",
            },
            Object {
              "label": "San Francisco AIDS Foundation",
              "value": "San Francisco AIDS Foundation",
            },
            Object {
              "label": "Section 8 HCV (tenant) voucher",
              "value": "Section 8 HCV (tenant) voucher",
            },
            Object {
              "label": "Self-Help for the Elderly",
              "value": "Self-Help for the Elderly",
            },
            Object {
              "label": "VASH voucher",
              "value": "VASH voucher",
            },
            Object {
              "label": "Other",
              "value": "Other",
            },
          ]
        }
        validation={[Function]}
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <Component>
      <CurrencyField
        fieldName="rental_assistances.0.assistance_amount"
        id="assistance_amount"
        label="Assistance Amount"
        validation={[Function]}
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <Component>
      <YesNoRadioGroup
        className="rental-assistance-recurring"
        falseValue="No"
        fieldName="rental_assistances.0.recurring_assistance"
        label="Recurring Assistance"
        trueValue="Yes"
        uniqId="new"
      />
    </Component>
  </Component>
  <Component
    expand={true}
  >
    <div
      className="form-grid_item column"
    >
      <Button
        classes="primary margin-right"
        disabled={false}
        id="rental-assistance-save"
        noBottomMargin={true}
        onClick={[Function]}
        text="Save"
        tiny={true}
      />
      <Button
        classes="secondary"
        disabled={false}
        id="rental-assistance-cancel"
        noBottomMargin={true}
        onClick={[MockFunction]}
        text="Cancel"
        tiny={true}
      />
    </div>
  </Component>
</InlineModal>
`;
